### ISSUE TITLE
How do I interpret loss while training?

### ISSUE THREAD

**Issue by lowcaloriesyrup** [tdrussell/diffusion-pipe#179]
Loss jumps around quite a bit while training a lora for Wan or Hunyuan using 1024px images. Why is this and how should I interpret what sufficiently small loss looks like in order to get a hint that the lora is 'done'.

```
[2025-03-28 19:56:05,234] [INFO] [logging.py:128:log_dist] [Rank 0] step=78, skipped=0, lr=[1.5644e-05], mom=[0.0]
steps: 78 loss: 0.1150 iter time (s): 18.305 samples/sec: 0.219
[2025-03-28 19:56:23,447] [INFO] [logging.py:128:log_dist] [Rank 0] step=79, skipped=0, lr=[1.5842000000000002e-05], mom=[0.0]
steps: 79 loss: 0.0814 iter time (s): 18.202 samples/sec: 0.220
[2025-03-28 19:56:41,782] [INFO] [logging.py:128:log_dist] [Rank 0] step=80, skipped=0, lr=[1.604e-05], mom=[0.0]
steps: 80 loss: 0.0995 iter time (s): 18.323 samples/sec: 0.218
[2025-03-28 19:56:59,935] [INFO] [logging.py:128:log_dist] [Rank 0] step=81, skipped=0, lr=[1.6238e-05], mom=[0.0]
steps: 81 loss: 0.1480 iter time (s): 18.141 samples/sec: 0.220
[2025-03-28 19:57:18,190] [INFO] [logging.py:128:log_dist] [Rank 0] step=82, skipped=0, lr=[1.6435999999999996e-05], mom=[0.0]
steps: 82 loss: 0.0812 iter time (s): 18.243 samples/sec: 0.219
[2025-03-28 19:57:36,344] [INFO] [logging.py:128:log_dist] [Rank 0] step=83, skipped=0, lr=[1.6633999999999997e-05], mom=[0.0]
steps: 83 loss: 0.0765 iter time (s): 18.142 samples/sec: 0.220
[2025-03-28 19:57:54,699] [INFO] [logging.py:128:log_dist] [Rank 0] step=84, skipped=0, lr=[1.6831999999999997e-05], mom=[0.0]
steps: 84 loss: 0.0668 iter time (s): 18.344 samples/sec: 0.218
[2025-03-28 19:58:13,012] [INFO] [logging.py:128:log_dist] [Rank 0] step=85, skipped=0, lr=[1.7029999999999998e-05], mom=[0.0]
steps: 85 loss: 0.0986 iter time (s): 18.302 samples/sec: 0.219
[2025-03-28 19:58:31,751] [INFO] [logging.py:128:log_dist] [Rank 0] step=86, skipped=0, lr=[1.7228e-05], mom=[0.0]
steps: 86 loss: 0.0541 iter time (s): 18.727 samples/sec: 0.214
```

**Comment by HDANILO** [comment]
I'm having the same issue, I made a dataset with 220 images, all captioned using florence2 detailed caption, it seems like I'm getting progress but now my LR is pretty much the same, and the loss is kinda all over the place(but doesn't increase)

```
[2025-03-28 21:13:45,789] [INFO] [logging.py:107:log_dist] [Rank 0] step=4827, skipped=0, lr=[2e-05], mom=[0.0]
steps: 4827 loss: 0.0545 iter time (s): 3.550 samples/sec: 0.282
[2025-03-28 21:13:49,229] [INFO] [logging.py:107:log_dist] [Rank 0] step=4828, skipped=0, lr=[2e-05], mom=[0.0]
steps: 4828 loss: 0.0670 iter time (s): 3.432 samples/sec: 0.291
[2025-03-28 21:13:52,667] [INFO] [logging.py:107:log_dist] [Rank 0] step=4829, skipped=0, lr=[2e-05], mom=[0.0]
steps: 4829 loss: 0.0556 iter time (s): 3.431 samples/sec: 0.291
[2025-03-28 21:13:56,109] [INFO] [logging.py:107:log_dist] [Rank 0] step=4830, skipped=0, lr=[2e-05], mom=[0.0]
steps: 4830 loss: 0.0884 iter time (s): 3.434 samples/sec: 0.291
[2025-03-28 21:13:59,549] [INFO] [logging.py:107:log_dist] [Rank 0] step=4831, skipped=0, lr=[2e-05], mom=[0.0]
steps: 4831 loss: 0.0736 iter time (s): 3.432 samples/sec: 0.291
[2025-03-28 21:14:02,985] [INFO] [logging.py:107:log_dist] [Rank 0] step=4832, skipped=0, lr=[2e-05], mom=[0.0]
steps: 4832 loss: 0.0801 iter time (s): 3.428 samples/sec: 0.292
[2025-03-28 21:14:06,423] [INFO] [logging.py:107:log_dist] [Rank 0] step=4833, skipped=0, lr=[2e-05], mom=[0.0]
steps: 4833 loss: 0.1229 iter time (s): 3.430 samples/sec: 0.292
[2025-03-28 21:14:09,863] [INFO] [logging.py:107:log_dist] [Rank 0] step=4834, skipped=0, lr=[2e-05], mom=[0.0]
steps: 4834 loss: 0.2176 iter time (s): 3.432 samples/sec: 0.291
```

**Comment by alisson-anjos** [comment]
The learning rate will stabilize after the warmup steps, so that the training does not start with a very high learning rate right away. To analyze the loss, it is interesting to use an evaluation dataset that can be configured through the configuration files and use the tensorboard to see the progress.

**Comment by HDANILO** [comment]
Is there any material on this @alisson-anjos ? Would love to follow a guide

**Comment by tdrussell** [comment]
https://github.com/tdrussell/diffusion-pipe/blob/668c797d698b5a2ce57abe4300fd33fd9f09b8ad/examples/hunyuan_video.toml#L7

The example TOML file shows how to set up evaluation.

You can even make the evaluation dataset a small subset of your training data. It is not a held-out validation loss at that point, but it is a stabilized version of training loss that will make a smooth graph.

**Comment by HDANILO** [comment]
> [diffusion-pipe/examples/hunyuan_video.toml](https://github.com/tdrussell/diffusion-pipe/blob/668c797d698b5a2ce57abe4300fd33fd9f09b8ad/examples/hunyuan_video.toml#L7)
> 
> Line 7 in [668c797](/tdrussell/diffusion-pipe/commit/668c797d698b5a2ce57abe4300fd33fd9f09b8ad)
> 
>  # eval_datasets = [ 
> The example TOML file shows how to set up evaluation.
> 
> You can even make the evaluation dataset a small subset of your training data. It is not a held-out validation loss at that point, but it is a stabilized version of training loss that will make a smooth graph.

Thank you, will definitely dig on it!

**Comment by lowcaloriesyrup** [comment]
Thank you for the insights!

> [diffusion-pipe/examples/hunyuan_video.toml](https://github.com/tdrussell/diffusion-pipe/blob/668c797d698b5a2ce57abe4300fd33fd9f09b8ad/examples/hunyuan_video.toml#L7)
> 
> Line 7 in [668c797](/tdrussell/diffusion-pipe/commit/668c797d698b5a2ce57abe4300fd33fd9f09b8ad)
> 
>  # eval_datasets = [ 
> The example TOML file shows how to set up evaluation.
> 
> You can even make the evaluation dataset a small subset of your training data. It is not a held-out validation loss at that point, but it is a stabilized version of training loss that will make a smooth graph.
